### PR TITLE
Refactor: Enable TypeScript strict null checks

### PR DESCRIPTION
**Pull request by calebmer**
Enables TypeScript `strictNullChecks` for increased type safety guarantees. A longer term solution to https://github.com/apollostack/apollo-client/issues/800

- [x] Make sure all of the significant new logic is covered by tests
- [x] Rebase your changes on master so that they can be merged easily
- [x] Make sure all tests and linter rules pass
- [x] Update CHANGELOG.md with your change

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -11,7 +11,7 @@ Expect active development and potentially significant breaking changes in the `0
 - Fix bug where there could be store inconsistencies for two dependent optimistic updates [PR #1144](https://github.com/apollostack/apollo-client/pull/1144)
 - expose partial in ObservableQuery#currentResult [PR #1097](https://github.com/apollostack/apollo-client/pull/1097)
 - Calculate `loading` from `networkStatus`. [PR #1202](https://github.com/apollostack/apollo-client/pull/1202)
-- Fix typings error with `strictNullChecks` [PR #1188](https://github.com/apollostack/apollo-client/pull/1188)
+- Fix typings error with `strictNullChecks` and enable TypeScript strict null checking in source code. [PR #1188](https://github.com/apollostack/apollo-client/pull/1188) and [PR #1221](https://github.com/apollostack/apollo-client/pull/1221)
```
Comment (helfer) [inline CHANGELOG.md]: What's the rationale for putting this under 0.8 instead of vNEXT?

**Comment by helfer** [comment]
Also, we should merge this before we merged anything else, because otherwise we'll have a ton of conflicts.

File: CHANGELOG.md (hunk 1 above)
Comment (calebmer) [inline CHANGELOG.md]: I didn’t realize there was a release and I missed the `vNEXT` when resolving the merge conflicts 😉

**Comment by calebmer** [comment]
@helfer agreed. Not sure if it caught any bugs that would ever realistically happen (because if they would realistically happen we likely already fixed it). Of course, we will never know now once this gets merged 😉

For fun, here are some candidates for bug fixes:

- https://github.com/apollostack/apollo-client/pull/1221/files#diff-f28988ac35502178bef737788c3d4cd5R703
- https://github.com/apollostack/apollo-client/pull/1221/files#diff-f28988ac35502178bef737788c3d4cd5R719
- https://github.com/apollostack/apollo-client/pull/1221/files#diff-5c95a4c033b67603fcaed250ea78b487R24
- https://github.com/apollostack/apollo-client/pull/1221/files#diff-e98247e2f1d34c868216269f192b07faR138

**Comment by helfer** [comment]
Yeah, those don't look like super-likely candidates, but let's see.

**Pull request merged**